File: examples/typescript/autonomousWorkflow.ts
import { AgentClient, PolicyViolationError } from '../../sdks/typescript/src';

async function runAutonomousWorkflow() {
  const client = new AgentClient({
    baseUrl: 'http://localhost:8080',
    apiKey: 'your_api_key',
    agentId: 'writer_agent_42'
  });

  const task = "Generate high-resolution images for marketing campaign";
  const estimatedCost = 1.25;

  console.log(`Starting autonomous task: ${task}`);
  console.log(`Estimated cost: $${estimatedCost}`);

  try {
    const result = await client.spend({
      amount: estimatedCost,
      context: {
        action: 'image_generation',
        service: 'dalle-3',
        priority: 'premium'
      }
    });

    console.log(`Authorization success! TxID: ${result.transaction_id}`);
    console.log(`Selected Rail: ${result.rail}`);

    // Proceed with image generation...
    console.log("Generating images...");

  } catch (error) {
    if (error instanceof PolicyViolationError) {
      console.warn(`Action restricted by policy: ${error.message}`);
      console.info("Agent Action: Adjusting parameters to reduce cost.");
    } else {
      console.error("Workflow halted due to error:", error);
    }
  }
}

runAutonomousWorkflow();

File: examples/python/autonomous_workflow.py
import sys
import os
from decimal import Decimal

# Add sdk path for demonstration
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "../../sdks/python")))

from agent_money import AgentClient, PolicyViolationError

def run_autonomous_task():
    client = AgentClient(
        base_url="http://localhost:8080",
        api_key="your_api_key",
        agent_id="research_agent_001"
    )

    # 1. Evaluate task cost before execution
    task_description = "Summarize top 10 papers on AGI safety"
    estimated_tokens = 5000
    estimated_cost = Decimal("0.05")

    print(f"Autonomous check: Executing '{task_description}'")
    print(f"Estimated cost: ${estimated_cost}")

    try:
        # Pre-authorizing the spend
        result = client.spend(
            amount=estimated_cost,
            context={
                "task": "summarization",
                "estimated_tokens": estimated_tokens,
                "urgency": "high"
            }
        )
        print(f"Spend approved! Transaction ID: {result.transaction_id}")
        
        # Now execute the actual task logic...
        print("Executing task logic...")
        
    except PolicyViolationError as e:
        print(f"Policy Blocked Action: {str(e)}")
        print("Agent Strategy: Fallback to a cheaper model or skip the task.")
    except Exception as e:
        print(f"System Error: {str(e)}")

if __name__ == "__main__":
    run_autonomous_task()
